src/System.Private.ServiceModel/tools/CertificateGenerator/CertificateManager.cs

```diff
@@ -0,0 +1,445 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Security.Cryptography.X509Certificates;
+
+namespace WcfTestBridgeCommon
+{
+    // This class manages the adding and removal of certificates.
+    // It also handles informing http.sys of which certificate to associate with SSL ports.
+    public static class CertificateManager
+    {
+        private static object s_certificateLock = new object();
+
+        // Dictionary of certificates installed by CertificateManager
+        
+        // Keyed by the Subject of the certificate - there should be only one valid cert per endpoint
+        // Valid certs are shareable across endpoints 
+        private static Dictionary<string, CertificateCacheEntry> s_myCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
+
+        // Keyed by endpoint address - each endpoint can only configure one invalid service certificate,
+        // Each endpoint address can have a unique cert
+        private static Dictionary<string, CertificateCacheEntry> s_myInvalidCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
+
+        // Keyed by CA subject
+        private static Dictionary<string, CertificateCacheEntry> s_rootCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
+
+        // Keyed by port, value is cert thumbprint
+        private static Dictionary<int, string> s_sslPorts = new Dictionary<int, string>();
+        
+        // When we install certificates via CreateAndInstallMachineCertificates, put local cert here 
+        // for reference when we need to add the sslport references
+        private static X509Certificate2 s_localCertificate = null;
+
+        public static EventHandler BridgeConfigurationChanged;
+
+        // Any change to the Bridge resource folder uninstalls all the certificates
+        // installed by those resources.
+        public static void OnResourceFolderChanged(string oldFolder, string newFolder)
+        {
+            BridgeConfigurationChanged(null, new EventArgs());
```

Comment:
There is a lot of code in here from the Bridge.  Not blocking, but please make a note to remove it at a later date.